packages/libs/avector/src/lock.rs

```diff
@@ -0,0 +1,59 @@
+//! Implementation of a light-weight lock guard, which is guarded by an atomic boolean.
+//!
+//! Most efficient when contention is low, acquiring the lock is a single atomic swap, and releasing
+//! is just one more atomic swap.
+//!
+//! This is adapted from the function outlined in the [std](https://doc.rust-lang.org/std/sync/atomic/fn.fence.html) documentation
+
+use crate::sync::{fence, spin_loop, AtomicBool, Ordering};
+
+pub(crate) struct LockGuard<'a> {
+    lock: &'a AtomicLock,
+}
+
+impl Drop for LockGuard<'_> {
+    fn drop(&mut self) {
+        self.lock.unlock();
+    }
+}
+
+// A mutual exclusion primitive based on spinlock.
+pub(crate) struct AtomicLock {
```

Comment:
The logic *seems* fine but I'd be very **very** cautious about that!
This is the most critical part of the application and basically replicates a spin lock. Why don't we use [`lock_api`](https://docs.rs/lock_api/0.4.9/lock_api/) for this instead of reinventing the wheel?

Reply:
Not really reinventing the wheel, this is the example from https://doc.rust-lang.org/std/sync/atomic/#examples

Reply:
Not reinventing the wheel, this is taken from the official rust documentation: https://doc.rust-lang.org/std/sync/atomic/fn.fence.html#examples

Reply:
([6055f86](https://github.com/hashintel/hash/pull/1399/commits/6055f867cdca747dc74567e7867e1b1651024cda))